Enable asynchronous job submission in BigQuery Insert Job

phanikumv opened the pull request:
- Add nowait flag to the insert_job method
- When nowait is True, the execution won't wait till the job results are available.
- By default, the job execution will wait till job results are available.

cc @dstandish @kaxil 

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

phanikumv committed:
Enable asynchronous job submission in BigQuery Insert Job.

- Add is_async flag to the insert_job method
- When is_async is True, the execution won't wait till the job results are available.
- By default, the job execution will wait till job results are available.

phanikumv committed:
Enable asynchronous job submission in BigQuery Insert Job.

- Add is_async flag to the insert_job method
- When is_async is True, the execution won't wait till the job results are available.
- By default, the job execution will wait till job results are available.

phanikumv committed:
fixup! Enable asynchronous job submission in BigQuery Insert Job.

phanikumv committed:
fixup! Enable asynchronous job submission in BigQuery Insert Job.

dstandish commented on tests/providers/google/cloud/hooks/test_bigquery.py:
```diff
@@ -943,6 +943,35 @@ def test_dbapi_get_sqlalchemy_engine(self):
+            },
+            mock_client.return_value,
+        )
+        mock_query_job.from_api_repr.return_value._begin.assert_called_once()
```
what you think of parameterizing this on `is_async`

when true, _begin called, and result not called
when false, the reverse

phanikumv replied:
parameterized test_insert_job with is_async flag being passed using a pytest fixture with True and False.Since the test class _BigQueryBaseTestClass inherits from unittest.TestCase , the pytest.mark.parameterize does not work and hence updated the test class to not inherit from TestCase by renaming setupMethod to setup_method

dstandish reviewed (commented):
~@phanikumv I think we should return `{"job_id": self.job_id}` from execute so it could be used in a sensor. can  you do   that and add a test for that behavoir?~
~thanks~

josh-fell commented on airflow/providers/google/cloud/hooks/bigquery.py:
```diff
@@ -1498,6 +1498,7 @@ def insert_job(
         job_id: Optional[str] = None,
         project_id: Optional[str] = None,
         location: Optional[str] = None,
+        is_async: Optional[bool] = False,
```
```suggestion
        is_async: bool = False,
```

phanikumv replied:
is_async is now changed to use a bool

josh-fell commented on airflow/providers/google/cloud/hooks/bigquery.py:
```diff
@@ -1514,6 +1515,8 @@ def insert_job(
         :param project_id: Google Cloud Project where the job is running
         :param location: location the job is running
+        :param is_async: specify whether to insert job in asynchronous mode
+        :type is_async: bool
```
```suggestion
```
The `:type:` directives are no longer needed. Sphinx can now directly infer from the type annotations rather than the docstrings for the API documentation.

phanikumv replied:
this is fixed now

phanikumv committed:
Parameterize the test for insert_job

- Pass is_async flag through a pytest fixture with True or False values
- Modify test_insert_job method to accept is_async as parameter
- Change is_async flag from Optional[bool] to bool

phanikumv committed:
remove test_insert_job_async post parameterizing test_insert_job

dstandish commented on airflow/providers/google/cloud/hooks/bigquery.py:
```diff
@@ -1514,6 +1515,8 @@ def insert_job(
         :param project_id: Google Cloud Project where the job is running
         :param location: location the job is running
+        :param is_async: specify whether to insert job in asynchronous mode
+        :type is_async: bool
```
```suggestion
```

dstandish commented on airflow/providers/google/cloud/hooks/bigquery.py:
```diff
@@ -1498,6 +1498,7 @@ def insert_job(
         job_id: Optional[str] = None,
         project_id: Optional[str] = None,
         location: Optional[str] = None,
+        is_async: bool = False,
```
i think we should rename `is_async`... it doesn't seem quite right because you are not _describing_ something (i.e. whether it _is_ async or not), but rather _instructing_ the func whether to make the call in an async way or not (so maybe `use_async` would be better).

but there's a suggestion from @uranusjr to use `nowait` which i think is probably better because it avoids the word `async` which is a bit fraught.

phanikumv replied:
agreed , made changes in the code accordingly

dstandish reviewed (commented):
everything looks good except i would just ask  that you rename `is_async` as suggested in the other comment.

dstandish committed:
Update airflow/providers/google/cloud/hooks/bigquery.py

phanikumv committed:
Rename is_async to nowait for better readability

dstandish reviewed (approved):
lgtm
